Use "subs" and "sups" font features for typographic scripts

MDLC01 opened the pull request:
This changes the `typographic` parameter of `sub` and `super` to use the "subs" and "sups" font features instead of special Unicode characters that should not be used that way.[^1]

- Closes #633.
- Closes #2146 by definitely making it a font issue (before this PR, we used characters that were not necessarily meant to be used together to write text).
- Closes #5112.
- Related to #3046.

[^1]: From [Section 7.8.1 of the Unicode 16.0.0 Core Specification](https://www.unicode.org/versions/Unicode16.0.0/core-spec/chapter-7/#G24762): "Superscript modifier letters are intended for cases where the letters carry a specific meaning, as in phonetic transcription systems, and are not a substitute for generic styling mechanisms for superscripting of text, as for footnotes, mathematical and chemical expressions, and the like."

MDLC01 commented:
New Computer Modern, Roboto, Liberation, do not support "sups" even for digits. So `show footnote: set super(typographic: true)` is probably a bad idea.

MDLC01 commented:
> It might actually be good to keep using the superscript characters for numbers, they tend to look better IMO.

But as you noted later, they shouldn't really be used that way. The real solution is to use the font features (at least for footnotes, it should be the default imo), which does indeed look betters with fonts that support it (e.g., Libertinus Serif uses the same glyph for "²" and `sups` "2" I believe)

> The question being whether typst's output counts as "rich text"

I don't see any reason why it wouldn't. Quoting [Wikipedia](https://en.wikipedia.org/wiki/Formatted_text):

> In [computing](https://en.wikipedia.org/wiki/Computing), **formatted text**, **styled text**, or **rich text**, as opposed to [plain text](https://en.wikipedia.org/wiki/Plain_text), is [digital text](https://en.wikipedia.org/wiki/E-text) which has styling information beyond the minimum of semantic elements: colours, styles ([boldface](https://en.wikipedia.org/wiki/Boldface), [italic](https://en.wikipedia.org/wiki/Italic_type)), [sizes](https://en.wikipedia.org/wiki/Point_(typography)), and special features in [HTML](https://en.wikipedia.org/wiki/HTML) (such as [hyperlinks](https://en.wikipedia.org/wiki/Hyperlink)).

PgBiel commented on crates/typst-library/src/text/shift.rs:
```diff
@@ -17,17 +14,18 @@ use crate::World;
-    /// Whether to prefer the dedicated subscript characters of the font.
+    /// Whether to use the `subs` OpenType feature to render the glyphs.
+    ///
+    /// Do not use this is the body contains non-text elements.
```
(Also: why not?)
```suggestion
    /// Do not use this if the body contains non-text elements.
```

MDLC01 replied:
Because the baseline will not be raised, and therefore non-text elements will appear as if they were not in a `super`/`sub` element. I adjusted the documentation to explain that.

PgBiel commented on crates/typst-library/src/text/shift.rs:
```diff
@@ -53,11 +51,15 @@ impl Show for Packed<SubElem> {
-                if is_shapable(engine, &text, styles) {
-                    return Ok(TextElem::packed(text));
-                }
+            if !is_text(&body) {
+                engine.sink.warn(warning!(
+                    self.span(),
+                    "typographic subscript element contains non-textual content";
+                    hint: "this may have undesired behavior";
+                    hint: "set the `typographic` parameter to `false`"
+                ))
             }
```
Not sure whether I like this, seems to be a bit too broad of a warning, and either way the output could be desired. We should, at the very least, add a pathological case to testing to verify whether it's worth adding this "soft-restriction".

MDLC01 replied:
I think the real solution would be to just make it work as expected, but we can't just change the baseline, because we don't want to change it for text elements as this is already handled by the font. I'm unsure how to change the baseline only for non-text elements. Additionally, we would need to get the information of the new baseline from the font, which I'm unsure is possible.

khaledhosny replied:
The OS/2 table has information on subscript/superscript position and scale, but unfortunately almost no application uses it, so fonts tend to hack junk values there (often auto generated by font editors).

MDLC01 replied:
This is annoying, compared to having no value at all...

MDLC01 replied:
All the fonts bundled in Typst have sensible metrics for `y{Sub,Super}scriptYSize` and `y{Sub,Super}scriptYOffset` (including New Computer Modern Math, which is not really intended to be used as a text font), so at least things will work well by default.

Below is the result of a prototype that synthesizes superscripts and subscripts for each bundled font by rescaling the glyphs according to `y{Sub,Super}scriptYSize` and shifting them according to `y{Sub,Super}scriptYOffset`. The font order is: DejaVu Sans Mono (4 variants), Libertinus Serif (6 variants), New Computer Modern (4 variants), and New Computer Modern Math (3 variants).

![Superscripts and subscripts appear in sensible places and in a sensible size.](https://github.com/user-attachments/assets/dd990956-e4d0-4e7a-a682-dac952dd7fa6)

However, there seems to be an error in the way `y{Sub,Super}scriptXOffset` (mainly used for italic correction) are encoded in DejaVu Sans Mono: in this font, the sign is inverted, meaning the offset is encoded as an offset to the left instead of as an offset to the right. The specification[^1][^2] isn't very precise, but the way I read it, it should be an offset to the right (or at least in the direction of the text direction): "from the glyph origin to the glyph origin of the subscript’s glyph". Below is what applying `y{Sub,Super}scriptXOffset` as an offset to the right looks like (I would recommend downloading the two images and switching between them to see the difference):

![image](https://github.com/user-attachments/assets/b0c48ed5-699e-4994-8987-483df9e83f7e)

Note that this works as expected for Libertinus Serif and New Computer Modern.

[^1]: https://learn.microsoft.com/en-us/typography/opentype/spec/os2#ysubscriptxoffset
[^2]: https://learn.microsoft.com/en-us/typography/opentype/spec/os2#ysuperscriptxoffset

PgBiel commented on crates/typst-library/src/text/shift.rs:
```diff
@@ -76,17 +78,18 @@ impl Show for Packed<SubElem> {
-    /// If this is enabled, Typst first tries to transform the text to
-    /// superscript codepoints. If that fails, it falls back to rendering
-    /// raised and shrunk normal letters.
+    /// Do not use this is the body contains non-text elements.
```
```suggestion
    /// Do not use this if the body contains non-text elements.
```

PgBiel reviewed (commented):
I'm also concerned with the font support deal, but leaving that discussion aside for a moment, I'll add some other comments.

PgBiel commented on crates/typst-library/src/text/shift.rs:
```diff
@@ -125,90 +132,10 @@ impl Show for Packed<SuperElem> {
-        ' ' => Some(' '),
-        _ => None,
-    }
+fn is_text(content: &Content) -> bool {
+    content.is::<SpaceElem>()
+        || content.is::<TextElem>()
+        || content
+            .to_packed::<SequenceElem>()
+            .is_some_and(|sequence| sequence.children.iter().all(is_text))
 }
```
This is also too restrictive and would also trigger on metadata, horizontal spacing and so on. I believe this complements my other comment on the warning.

MDLC01 replied:
On a side note, `h` elements might work slightly unexpectedly: since the font size does not change when `typographic: true`, `1em` in `super` or `sub` with `typographic: true` is the same as `1em` outside. This means `h(1em)` will appear larger with `typographic: true` than with `typographic: false`.

khaledhosny commented:
CSS [specification](https://drafts.csswg.org/css-fonts/#font-variant-position-prop) describe a fallback behavior that should help with inconsistent font coverage of these features. Quoting some of that text:
> Because of the semantic nature of subscripts and superscripts, when the value is either sub or super for a given contiguous run of text, if a variant glyph is not available for all the characters in the run, simulated glyphs should be synthesized for all characters using reduced forms of the glyphs that would be used without this feature applied.

MDLC01 commented:
Just to keep you updated, I am now working on this again, and I hope to have a working solution in the upcoming days/weeks. Feel free to close the PR to prevent clutter.

laurmaedje commented:
If you plan to work on it in the next days or weeks, there is no need to close.

MDLC01 commented on crates/typst-library/src/text/shift.rs:
```diff
@@ -109,106 +120,88 @@ pub struct SuperElem {
     let world = engine.world;
-    for family in TextElem::font_in(styles) {
-        if let Some(font) = world
+    let text = body.plain_text();
```
I'm not sure this is the right way to get the inner text to test for font support.

MDLC01 replied:
We no longer do that

MDLC01 commented on tests/ref/super-underline.png:
Ideally, `underline(super(..))` should have the underline based on the outer baseline, and `super(underline(..))` should be based on the script baseline, regardless of whether the scripts are synthesized or not.

It did not work like that before (in fact, I think this PR does not change the behavior in this regard), so this can also be left as a future improvement.

MDLC01 commented on tests/ref/super-1em.png:
For now, `1em` within a script resolves to the same absolute length as `1em` outside. Maybe this should be changed to account for the reduced script size.

MDLC01 commented on tests/suite/text/shift.typ:
```diff
@@ -8,12 +8,72 @@
+  [x#sub[1]], [x#sub[5n]], [x#sub[2 #box(square(size: 6pt))]],
+)
+
+--- sub-super-synthesized ---
 #set super(typographic: false, baseline: -0.25em, size: 0.7em)
 n#super[1], n#sub[2], ... n#super[N]
```
This test (previously named `sub-super-non-typographic`) has slightly changed for reasons I can't figure out.

MDLC01 reviewed (commented):
I pushed a prototype that uses the OpenType features by default, and synthesizes scripts when `typographic` is set to `false`.

When the OpenType features are used, non-textual elements are raised as well using the metrics provided by the font (although italic compensation, i.e., horizontal offset, is not applied to. When scripts are synthesized, the metrics from the fonts are used (for if they are available, and otherwise, there is a fallback to the metrics that were used before this PR.

This is still not fully ready to be merged. Below are some concerns and open questions.

Currently, the `typographic` argument is a boolean. When set to `false`, scripts are synthesized. When set to `true`, the corresponding OpenType feature is enabled, and that's it. There is no check that the font does in fact support the feature, and that it has subs/sups glyphs for all the characters used. I don't think there is an easy way to do that, other than manually comparing that all glyphs are different. What's bad is that the default text font, Libertinus Serif, does not have many sups/subs glyphs, so users are likely to encounter an issue if we set `typographic: true` by default. This is similar to the way `smallcaps` currently works, except Libertinus Serif has better support for `smcp` and `c2sc` than for `subs` and `sups`.

Finally, fonts can provide a different scaling for both axes for scripts (typically, the glyphs are scaled down vertically more than horizontally). When synthesizing scripts and when scaling non-textual elements, we scale them evenly in both axes, according to the font's vertical scaling factor. Should we use the horizontal scaling factor instead? Or should we use both?

MDLC01 commented on tests/ref/super-highlight.png:
Similar issue here as with underlines: https://github.com/typst/typst/pull/5777/#discussion_r2016787652.

laurmaedje commented:
>  There is no check that the font does in fact support the feature, and that it has subs/sups glyphs for all the characters used. I don't think there is an easy way to do that, other than manually comparing that all glyphs are different.

For a basic check that the feature is supported, you could probably check for the [GSUB table](https://docs.rs/ttf-parser/latest/ttf_parser/struct.FaceTables.html#structfield.gsub), iterate its [features](https://docs.rs/ttf-parser/latest/ttf_parser/opentype_layout/struct.LayoutTable.html#structfield.features) and check that it has matches for the `subs` and `sups` [Tags](https://docs.rs/ttf-parser/latest/ttf_parser/struct.Tag.html). I haven't tested it, but worth a try.

khaledhosny commented:
> > There is no check that the font does in fact support the feature, and that it has subs/sups glyphs for all the characters used. I don't think there is an easy way to do that, other than manually comparing that all glyphs are different.
> 
> For a basic check that the feature is supported, you could probably check for the [GSUB table](https://docs.rs/ttf-parser/latest/ttf_parser/struct.FaceTables.html#structfield.gsub), check thats its [features](https://docs.rs/ttf-parser/latest/ttf_parser/opentype_layout/struct.LayoutTable.html#structfield.features) [have matches](https://docs.rs/ttf-parser/latest/ttf_parser/opentype_layout/struct.RecordList.html#tymethod.find) for the `subs` and `sups` [Tags](https://docs.rs/ttf-parser/latest/ttf_parser/struct.Tag.html). I haven't tested it, but worth a try.

HarfBuzz has [`hb_ot_layout_lookup_collect_glyphs()`](https://harfbuzz.github.io/harfbuzz-hb-ot-layout.html#hb-ot-layout-lookup-collect-glyphs) which can be used to get the glyphs covered by a given lookup (and other APIs to get lookups for a feature). I don't know if rustybuzz has equivalent APIs.

MDLC01 commented:
Thanks for the pointer! I will look into that.

MDLC01 commented:
I now enable the feature only if it is supported (in the sense that all glyphs required to display a specific script are available). I still want to add more tests before merging to make sure everything works as intended, but it is pretty much in the final state for now.

The commits are a bit all over the place if you review the PR from the beginning. Tell me if you would like me to clean all that.

MDLC01 commented:
I cleaned up the commits and added some tests.

There is still the question of whether `1em` within scripts should resolve to the same absolute size as outside the scripts or not (https://github.com/typst/typst/pull/5777#discussion_r2016789648). For now, `1em` within scripts is the same as `1em` outside of scripts, as illustrated by the `super-1em` test. If we decide to keep this behavior, this is ready to be merged!

Additionally, the behavior of `underline` and `highlight` together with `super` and `sub` could be improved. I think this can be left as a future improvement, as the current behavior is already imperfect, and comparable to the behavior with this PR (https://github.com/typst/typst/pull/5777#discussion_r2016787652).

MDLC01 commented on tests/ref/sub-super-typographic.png:
The squares aren't perfectly aligned to the superscript/subscript baseline in this test. I believe this to be due to imperfect metrics in Libertinus Serif.

MDLC01 replied:
Seems to be https://github.com/alerque/libertinus/issues/511.

laurmaedje commented on crates/typst-library/src/text/font/mod.rs:
```diff
@@ -215,6 +215,10 @@ pub struct FontMetrics {
     /// Recommended metrics for an overline.
     pub overline: LineMetrics,
+    /// Metrics for subscripts.
+    pub subscript: ScriptMetrics,
```
I think it would be cleaner to have these as options instead of just setting them to a default and handling the actual defaults in shift.rs rather than on `ScriptMetrics`.

MDLC01 replied:
I mirrored the way this is done for line metrics. Should I change them as well?

laurmaedje replied:
We can take a look at changing line metrics separately.

laurmaedje commented on crates/typst-library/src/text/font/mod.rs:
```diff
@@ -283,6 +310,34 @@ pub struct LineMetrics {
+    pub vertical_offset: Em,
+}
+
+impl ScriptMetrics {
+    pub const fn default_with_vertical_offset(offset: Em) -> Self {
+        Self {
+            width: Em::new(0.6),
+            height: Em::new(0.6),
+            horizontal_offset: Em::zero(),
+            vertical_offset: offset,
+        }
+    }
+
+    pub const fn default_subscript() -> Self {
+        Self::default_with_vertical_offset(Em::new(-0.2))
+    }
+
+    pub const fn default_superscript() -> Self {
+        Self::default_with_vertical_offset(Em::new(0.5))
+    }
+}
```
this would then all be unified with the methods on `ScriptKind`. I think it's also cleaner to not have the defaults live with the element rather than the font parsing code.

laurmaedje commented on crates/typst-library/src/text/shift.rs:
```diff
@@ -108,107 +120,55 @@ pub struct SuperElem {
 
-    false
-}
+    pub const fn read_metrics(self, font_metrics: &FontMetrics) -> ScriptMetrics {
```
we could return `&ScriptMetrics` here and for the defaults have them in `static`s and return references to those.

laurmaedje commented on crates/typst-library/src/layout/length.rs:
```diff
@@ -91,6 +91,10 @@ impl Length {
         )
     }
+
+    pub fn to_em(&self, text_size: Abs) -> Em {
```
please move this above `ensure_...` and add a brief doc comment

MDLC01 replied:
Done

laurmaedje commented on crates/typst-layout/src/inline/shaping.rs:
```diff
@@ -789,6 +803,52 @@ fn shape_segment<'a>(
+                        font.rusty().glyph_index(c).is_some_and(|i| coverage.contains(i))
+                    })
+                    .then(|| {
+                        ctx.features.push(Feature::new(settings.kind.feature(), 1, ..));
```
this is dangerous as the context is reused across multiple invocations of shape_segment. maybe the logic can be appling for the whole `shape` run instead of in `shape_segment`? shape_segment's is separate for font fallback and I don't think how fonts fall back should affect the super/sub-scripts.

MDLC01 replied:
I'm still unsure why font fallback shouldn't affect scripts. For example, `text.cover` should affect shifts right? For example, if I'm writing a document in two scripts and I use different fonts for those, I don't want shifts to be affected only by whichever font comes first, regardless of its `cover`.

laurmaedje replied:
Maybe I'm misunderstanding what's happening here, but all I'm saying is that you are pushing into `ctx.features` for each individual shape segment. Let's say you have three segments (because we have font fallback twice). Once you've arrived at the third segment, `ctx.features` will contain the feature thrice, which doesn't seem right.

laurmaedje replied:
Yeah that should fix it. Alternatively, we could avoid the allocation by pushing the feature (but outside of the compute_synthesized_shift function) and then popping it again after the shape plan is created.

laurmaedje commented on crates/typst-layout/src/inline/shaping.rs:
```diff
@@ -789,6 +803,52 @@ fn shape_segment<'a>(
+    // transformed in any way, and so those values are neutral (`(0, 0, 1)`). If
+    // scripts should be synthesized, those values determine how to transform
+    // the rendered text to display scripts as expected.
+    let (script_shift, script_compensation, scale) = match script_settings {
```
could you try to extract this logic into a function. shape_segment is already very long.

MDLC01 replied:
I extracted this into a `compute_synthesized_shift` function. Let me know if you find a better name.

laurmaedje replied:
The name sounds fine to me. I would suggest pulling the match on the option out of the function and keeping the logic in the function for the case where we do have shift settings.

MDLC01 replied:
Done

laurmaedje commented on crates/typst-layout/src/inline/shaping.rs:
```diff
@@ -723,6 +736,7 @@ fn shape_segment<'a>(
     base: usize,
     text: &str,
     mut families: impl Iterator<Item = &'a FontFamily> + Clone,
+    script_settings: Option<ScriptSettings>,
```
could these be part of the context?

MDLC01 replied:
This is indeed better

laurmaedje commented on crates/typst-layout/src/inline/shaping.rs:
```diff
@@ -64,6 +65,8 @@ pub struct ShapedGlyph {
     /// The vertical offset of the glyph.
     pub y_offset: Em,
+    /// How much to scale the glyph compared to its normal size.
+    pub scale: Em,
```
it's a bit unfortunate to have both this and `size`. couldn't we instead move `size` here?

MDLC01 replied:
I will wait for an answer to https://github.com/typst/typst/pull/5777#discussion_r2160401965 before addressing that because moving prat of the logic to `shape` could make this easier.

MDLC01 commented on crates/typst-kit/src/package.rs:
I'm not sure what this is referring to, and I don't see any change to that file in GitHub's "Files changed" tab.

laurmaedje replied:
The comment was from March. Probably already fixed.

MDLC01 commented on tests/suite/text/shift.typ:
```diff
@@ -1,19 +1,78 @@
 
+--- sub-super-typographic ---
+#set text(size: 20pt)
+// Libertinus Serif supports "subs" and "sups" for `typo`, but not for `synth`.
```
Done

MDLC01 commented on tests/suite/text/shift.typ:
```diff
@@ -1,19 +1,78 @@
-  [x#super[1]], [x#super[5n]], [x#super[2 #box(square(size: 6pt))]],
-  [x#sub[1]], [x#sub[5n]], [x#sub[2 #box(square(size: 6pt))]],
+  [Typo.], [Fallb.], [Synth.],
+  [x#super[1 #sq]], [x#super[5: #sq]], [x#super(typographic: false)[2 #sq]],
```
This was indeed due to the space

MDLC01 commented on tests/suite/text/shift.typ:
```diff
@@ -1,19 +1,78 @@
-The claim#super[#underline[\[4\]]] has been disputed. \
-It really has been#super(box(text(baseline: 0pt, underline[\[4\]]))) \
+#set super(typographic: false)
+#underline[The claim#super[4]] has been disputed. \
+The claim#super[#underline[4]] has been disputed. \
+The claim #underline(super[4]) has been disputed. \
+#set super(typographic: true)
+#underline[The claim#super[4]] has been disputed. \
+The claim#super[#underline[4]] has been disputed. \
+The claim #underline(super[4]) has been disputed.
+
+--- super-highlight ---
+#set super(typographic: false)
+#highlight[The claim#super[4]] has been disputed. \
+The claim#super[#highlight[4]] has been disputed. \
+It really has been#super(highlight[4]) \
+#set super(typographic: true)
+#highlight[The claim#super[4]] has been disputed. \
+The claim#super[#highlight[4]] has been disputed. \
+It really has been#super(highlight[4])
```
Done

MDLC01 commented on tests/suite/text/shift.typ:
```diff
@@ -1,19 +1,78 @@
+It really has been#super(highlight[4])
+
+--- super-1em ---
+Test#super[#box(rect(height: 1em))]#box(rect(height: 1em))
```
Actually, the reference image was removed when updating the tests!

laurmaedje commented on crates/typst-library/src/text/mod.rs:
```diff
@@ -752,6 +752,11 @@ pub struct TextElem {
+    /// Whether subscripts or superscripts are enabled.
+    #[internal]
+    #[ghost]
+    pub subpercript: Option<ScriptSettings>,
```
That message was also old. I think I had a few pending review comments from an initial review attempt and these now showed up.

MDLC01 commented on tests/suite/text/shift.typ:
```diff
@@ -1,19 +1,78 @@
+|#sub(typographic: false)[longscript]|
+
+--- scripts-with-bundeled-fonts ---
+#let test(font, weights, styles) = {
```
I don't remember why I added this test, but it is indeed useful to make sure we use the font-provided metrics in the right way (e.g., no sign error). Maybe I could make the test shorter by not testing all fonts, since we only need to make sure we use the metrics properly once. From what I remember, I simply copied the code from https://github.com/typst/typst/pull/5777#discussion_r2003568010 because I figured it could be a useful test.

I added a comment and also set `typographic: false` for both `super` and `sub`, in order to make sure we actually synthesize the scripts. The reference image does not change because none of the bundled fonts provide script glyphs for X and x.

MDLC01 commented on crates/typst-library/src/text/shift.rs:
```diff
@@ -108,107 +120,55 @@ pub struct SuperElem {
-        _ => None,
-    }
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
+pub struct ScriptSettings {
```
Done

MDLC01 commented on crates/typst-library/src/text/font/mod.rs:
```diff
@@ -283,6 +310,34 @@ pub struct LineMetrics {
+/// Metrics for subscripts or superscripts.
+#[derive(Debug, Copy, Clone)]
+pub struct ScriptMetrics {
+    pub width: Em,
```
Done

MDLC01 commented on crates/typst-library/src/text/shift.rs:
```diff
@@ -17,11 +16,17 @@ use crate::World;
 #[elem(title = "Subscript", Show)]
 pub struct SubElem {
-    /// Whether to prefer the dedicated subscript characters of the font.
+    /// Whether to create artificial subscripts by shifting and scaling down
+    /// regular glyphs.
```
I meant "shifting _(implicitly down)_ and scaling down", but I agree this sounds weird.

I changed it to "lowering and ..." for subscripts and "raising and ..." for superscripts.

MDLC01 commented on crates/typst-library/src/text/shift.rs:
```diff
@@ -17,11 +16,17 @@ use crate::World;
+    /// Whether to create artificial subscripts by shifting and scaling down
+    /// regular glyphs.
+    ///
+    /// Ideally, subscripts glyphs should be provided by the font (using the
+    /// `subs` OpenType feature). Otherwise, Typst is able to synthesize
+    /// subscripts as explained above.
```
Done

MDLC01 commented on crates/typst-library/src/text/shift.rs:
```diff
@@ -30,17 +35,19 @@ pub struct SubElem {
+    ///
+    /// This only applies to synthesized subscripts. In other words, this has no
+    /// effect if `typographic` is `{true}` and the font provides the necessary
+    /// subscript glyphs.
```
I added:

> If set to `{auto}`, the baseline is shifted according to the metrics provided by the font.

Should I also mention the default values in case the font does not provide said metrics?

Regarding the direction the baseline shift is applied, this is not the only place where this problem occurs in Typst documentation (e.g., [`box.baseline`](https://typst.app/docs/reference/layout/box/#parameters-baseline) has the same issue), but I will only change it here for now.

For both superscripts and subscripts, the baseline shift applies downward. This means, for superscripts, users need to provide a negative value. This is consistent with `sub` and `box`.

laurmaedje replied:
> Should I also mention the default values in case the font does not provide said metrics?

Yeah, I think that would be good.

MDLC01 replied:
Done

MDLC01 commented on crates/typst-library/src/text/shift.rs:
```diff
@@ -30,17 +35,19 @@ pub struct SubElem {
+    /// This only applies to synthesized subscripts. In other words, this has no
+    /// effect if `typographic` is `{true}` and the font provides the necessary
+    /// subscript glyphs.
+    pub size: Smart<TextSize>,
```
Done

MDLC01 commented on crates/typst-library/src/text/shift.rs:
```diff
@@ -108,107 +120,55 @@ pub struct SuperElem {
-            .styled(TextElem::set_baseline(self.baseline(styles)))
-            .styled(TextElem::set_size(self.size(styles))))
+    fn show(&self, _: &mut Engine, styles: StyleChain) -> SourceResult<Content> {
+        let outer_text_size = TextElem::size_in(styles);
```
Done

MDLC01 commented:
I will address the remaining comments later, as they require more thinking.

MDLC01 commented:
I have now implemented or asked for clarification on the remaining suggestions.

MDLC01 commented:
This is now ready for review. As noted in https://github.com/typst/typst/pull/5777#discussion_r2173765801, I'm unsure in which places `TextElem::size_in(shaped_text.styles)` vs. `glyph.size` should be used.

laurmaedje commented on crates/typst-layout/src/inline/shaping.rs:
```diff
@@ -932,6 +961,65 @@ fn shape_segment<'a>(
     ctx.used.pop();
 }
 
+/// Returns a `(script_shift, script_compensation, scale)` triplet describing
```
The doc comment is outdated. The function name doesn't seem 100% correct anymore since it also produces the feature. How about `determine_shift`? Or splitting it into two functions `compute_synthesized_shift(font: &Font, settings: ShiftSettings) -> (Em, Em, Em)` and `shift_feature(text: &str, font: &Font, settings: ShiftSettings) -> Option<Feature>` and then tying it together above.

laurmaedje commented:
Thanks a lot for your work and patience with this!

khaledhosny commented on crates/typst-layout/src/inline/shaping.rs:
```diff
@@ -932,6 +961,64 @@ fn shape_segment<'a>(
+            // In case of a problem, we just early exit
+            let gsub = font.rusty().tables().gsub?;
+            let subtable_index =
+                gsub.features.find(settings.kind.feature())?.lookup_indices.get(0)?;
```
If I’m reading this correctly, this code collects the glyph coverage from the first lookup in the feature, but the feature can have multiple lookups. This might be the cause for #7249.

MDLC01 replied:
So if I understand correctly, I should instead consider the union of the coverages from all subtables from all lookups to test whether there exists a scripted glyph?

khaledhosny commented on crates/typst-layout/src/inline/shaping.rs:
```diff
@@ -932,6 +961,64 @@ fn shape_segment<'a>(
+                .lookups
+                .get(subtable_index)?
+                .subtables
+                .get::<SubstitutionSubtable>(0)?
```
This also gets the coverage from the first subtable in the lookup, but lookups can have more than one subtable.
